extend config to show addOuput (#6963)
aws-amplify/docs: src/pages/gen2/reference/amplifyconfiguration/index.mdx

You can also manually create this file for a specified Amplify app ID and branch, or an AWS CloudFormation stack name with [`amplify generate config`](/gen2/reference/cli-commands#amplify-generate-config).

{/* @TODO full type/interface to reference */}

## Extending Amplify configuration file

> The `amplifyconfiguration.json` file is not just a static artifact; it's designed to be extendable to suit the evolving needs of your application. By leveraging the `addOutput` method you can programmatically add or modify configurations.
> This is particularly useful for customizing outputs that are not directly exposed through the Amplify constructs or for dynamically adjusting your app's configuration in response to changes in your backend strategy.

Reviewer: ```suggestion
The `amplifyconfiguration.json` file is not just a static artifact; it's designed to be extendable to suit the evolving needs of your application. By leveraging the `addOutput` method you can programmatically add or modify configurations. This is particularly useful for customizing outputs that are not directly exposed through the Amplify constructs or for dynamically adjusting your app's configuration in response to changes in your backend strategy.
```

was this an intentional line break?

Author: yes, for readability but can change it to combined.

Reviewer: ah if it renders as two separate paragraphs can we separate in the markdown with a newline for readability?

Author: keeping it as one paragraph should be fine.

One common scenario where extending the configuration becomes handy is when you need to add custom outputs or overwrite existing configurations without manual file edits.

### Override configuration

Consider a scenario where you want to override output parameters in your `amplifyconfiguration.json` that specify an S3 bucket and its region that your application will use for storing files.

> ```ts title="amplify/backend.ts"

Reviewer: can we show the imports as well?

const backend = defineBackend({
  auth, 
  data, 
});

backend.addOutput({
  aws_user_files_s3_bucket: "my-externally-managed-bucket",
  aws_user_files_s3_bucket_region: "us-east-1",
});
```

### Custom configuration

> In addition to overriding existing configurations, you can also add custom output parameters to your `amplifyconfiguration.json`. This is useful for including unique identifiers, custom flags, or any other information that might be necessary for your application's logic or configuration.

Reviewer: ```suggestion
In addition to overriding existing configurations, you can also add custom output parameters to your `amplifyconfiguration.json`. This is useful for surfacing arbitrary outputs, values from custom CDK resources, or any other information that might be necessary for your application's logic or configuration.
```

Reviewer: adding a callout for surfacing CDK values

You can also manually create this file for a specified Amplify app ID and branch, or an AWS CloudFormation stack name with [`amplify generate config`](/gen2/reference/cli-commands#amplify-generate-config).

{/* @TODO full type/interface to reference */}

## Extending Amplify configuration file

> The `amplifyconfiguration.json` file is not just a static artifact; it's designed to be extendable to suit the evolving needs of your application. By leveraging the `addOutput` method you can programmatically add or modify configurations. This is particularly useful for customizing outputs that are not directly exposed through the Amplify constructs or for dynamically adjusting your app's configuration in response to changes in your backend strategy.

Reviewer 2: 1. How will customers find what a `addOutput` method is?
2. Do we allow to modify existing configurations? See https://github.com/aws-amplify/amplify-backend/pull/978#discussion_r1479750371

Author: modified the wording to callout a this is a method on the backend and removed override
added a callout about modifying amplify managed resources

One common scenario where extending the configuration becomes handy is when you need to add custom outputs or overwrite existing configurations without manual file edits.

### Override configuration

Consider a scenario where you want to override output parameters in your `amplifyconfiguration.json` that specify an S3 bucket and its region that your application will use for storing files.

```ts title="amplify/backend.ts"
import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";

const backend = defineBackend({
  auth, 
  data, 
});

backend.addOutput({
  aws_user_files_s3_bucket: "my-externally-managed-bucket",
  aws_user_files_s3_bucket_region: "us-east-1",
});
```

### Custom configuration

> In addition to overriding existing configurations, you can also add custom output parameters to your `amplifyconfiguration.json`. This is useful for surfacing arbitrary outputs, values from custom CDK resources, or any other information that might be necessary for your application's logic or configuration.

Reviewer 2: Currently libraries don't support passing or exposing the "custom" configuration part of the `amplifyconfiguration.json` to developers. How are they expected to use these custom configurations in their frontend app?

Author: added an example on using this in the `amplify.configure()`.

once the custom examples are out will add the links

```ts title="amplify/backend.ts"
import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";

const backend = defineBackend({
  auth, 
  data, 
});

backend.addOutput({
  custom: {
    "message": "Hello World",
  }

> });

Reviewer 2: Perhaps better to show a "custom" cdk resource here rather than a hello world message.

Author: The custom CDK examples will be part of different PR. reference: https://github.com/aws-amplify/docs/pull/6901

You can also manually create this file for a specified Amplify app ID and branch, or an AWS CloudFormation stack name with [`amplify generate config`](/gen2/reference/cli-commands#amplify-generate-config).

{/* @TODO full type/interface to reference */}

## Extending Amplify configuration file

The `amplifyconfiguration.json` file is not just a static artifact; it's designed to be extendable to suit the evolving needs of your application. By leveraging the `addOutput` method from your `backend`, you can programmatically add configurations. This is particularly useful for customizing outputs that are not directly exposed through the Amplify constructs or for dynamically adjusting your app's configuration in response to changes in your backend strategy.

<Callout warning>

> Overriding Amplify managed configurations on `amplifyconfiguration.json` is not supported.

Reviewer: ```suggestion
Overriding Amplify-managed configurations on `amplifyconfiguration.json` is not supported.
```
